core/libp2p/libp2p.hpp

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_LIBP2P_HPP
+#define KAGOME_LIBP2P_HPP
+
+#include <memory>
+
+#include "common/result.hpp"
+#include "libp2p/common/peer_info.hpp"
+#include "libp2p/connection/connection.hpp"
+#include "libp2p/error/error.hpp"
+#include "libp2p/multi/multiaddress.hpp"
+#include "libp2p/multi/multihash.hpp"
+#include "libp2p/routing/router.hpp"
+#include "libp2p/store/record_store.hpp"
+#include "libp2p/swarm/swarm.hpp"
+
+namespace libp2p {
+  /**
+   * Provides a way to instantiate a libp2p instance via joining different
+   * components to it and dial to the peer
+   */
+  class Libp2p {
+    /**
+     * Add a swarm component
+     * @param swarm to be added
+     */
+    virtual void addSwarm(std::unique_ptr<swarm::Swarm> swarm) = 0;
+
+    /**
+     * Add a routing component
+     * @param routing to be added
+     */
+    virtual void addPeerRouting(std::unique_ptr<routing::Router> router) = 0;
+
+    /**
+     * Add a record store component
+     * @param record store to be added
+     */
+    virtual void addRecordStore(std::unique_ptr<store::RecordStore> store) = 0;
+
+    using DialResult =
```

Comment:
type alias can be common in folder, I guess

Reply:
it is used only here

Reply:
[One](https://github.com/soramitsu/polkadot/pull/9/files/93e7d7e3ef153e608add51acf1dac2af835de9e7#diff-8fbd3ed430b5384639fe3023ec028d42R32), [two](https://github.com/soramitsu/polkadot/pull/9/files/93e7d7e3ef153e608add51acf1dac2af835de9e7#diff-8fbd3ed430b5384639fe3023ec028d42R43), [three](https://github.com/soramitsu/polkadot/pull/9/files/93e7d7e3ef153e608add51acf1dac2af835de9e7#diff-dbc94a507dd8ffe3afcfbaf021e2fbc2R29)

Reply:
We discussed with @Warchant such a possibility and agreed that "common" file is going to introduce a lot of problems with `include`-s in that file. For example, now we need only a `Result<Connection, Error>`, but later we will add some kind of `variant<Class1, Class2>` into the same file, which will lead to uneccessary dependencies in the old files.